aioxmpp/pep/service.py

```diff
@@ -0,0 +1,238 @@
+            )
+        if register_feature:
+            self._disco_server.register_feature(node_namespace)
+        if notify:
+            self._disco_server.register_feature(node_namespace + "+notify")
+        self._pep_node_claims[node_namespace] = handler, register_feature, notify
+
+    def unclaim_pep_node(self, node_namespace):
+        """
+        Unclaim `node_namespace`.
+
+        The feature for `node_namespace` and the ``+notify`` feature
+        are automatically retracted if they were set by
+        :method:`claim_pep_node`.
+
+        :param node_namespace: The PubSub node whose handler shall be unset.
+
+        :raises KeyError: If the no handler is registered for
+            `node_namespace`.
+        """
+        _, feature, notify = self._pep_node_claims.pop(node_namespace)
+        if notify:
+            self._disco_server.unregister_feature(node_namespace + "+notify")
+        if feature:
+            self._disco_server.unregister_feature(node_namespace)
+
+    @asyncio.coroutine
+    def _check_for_pep(self):
+        # XXX: should this be done when the stream connects
+        # and we use the cached result later on (i.e. disable
+        # the PEP service if the server does not support PEP)
+        disco_info = yield from self._disco_client.query_info(
+            self.client.local_jid.bare()
+        )
+
+        for item in disco_info.identities.filter(attrs={"category": "pubsub"}):
+            if item.type_ == "pep":
+                break
+        else:
+            raise RuntimeError("server does not support PEP")
+
+    @service.depsignal(aioxmpp.PubSubClient, "on_item_published")
+    def _handle_pubsub_publish(self, jid, node, item, *, message=None):
+        try:
+            handler, _, _ = self._pep_node_claims[node]
+        except KeyError:
+            return
+
+        # TODO: handle empty payloads due to (mis-)configuration of
+        # the node specially.
+        handler(jid, node, item, message=message)
+
+    def publish(self, node, data, *, id_=None):
+        """
+        Publish an item `data` in the PubSub node `node` on the
+        PEP service associated with the user's JID.
+
+        If no `id_` is given it is generated by the server (and may be
+        returned).
+
+        :param node: The PubSub node to publish to.
+        :param data: The item to publish.
+        :type data: An XSO representing the paylaod.
+        :param id_: The id the published item shall have.
+
+        :returns: The PubSub id of the published item or
+            :data:`None` if it is unknown.
+
+        """
+        yield from self._check_for_pep()
+        return (yield from self._pubsub.publish(None, node, data, id_=id_))
+
+    def subscribe(self, jid, node):
+        yield from self._pubsub.subscribe(jid, node)
+
+class RegisteredPEPNode:
+    # XXX: do we want to provide metadata in this class?
+    on_event_received = callbacks.Signal()
```

Comment:
I’d at least add a ``notify`` property where notification can be turned on and off. Maybe even enabling/disabling the exporting of the feature should be possible.

I would even go that far that the RegisteredPEPNode should be the object used within the PEPService and all normal interfacing which is done through PEPService for enabling/disabling notify etc. should simply be a forward to the RegisteredPEPNode. It should also be possible to obtain the RegisteredPEPNode when claiming a namespace so that callbacks can be used without the descriptor (useful if you implement a PEP application in the application layer, without services).

Using the descriptor then is just a simple interface to obtain the RegisteredPEPNode.